### ISSUE TITLE
Take pod number into consideration for scheduling algorithm in `priority` stage

### ISSUE THREAD

**Issue by dalanlan** [kubernetes/kubernetes#15531]
Currently we only consider cpu and memory resource (being requested) when calculating _priority_ score for each node in `LeastRequestedPriority`.
Seems that pod number should be a factor to consider during scheduling since it has the upper bound set by kubelet. 
Is there any particular reason to avoid this?
/cc @davidopp

**Comment by davidopp** [comment]
Well, there are two issues
(1) currently, scheduler does not have a predicate function to avoid putting too many pods on the node
(2) scheduler also does not bias away from nodes with large number of pods

I don't think we need (2) if we have (1).

It would be good to implement (1). If you want to pick up the predicate function changes from
https://github.com/kubernetes/kubernetes/pull/11161
and create a new PR, that would be great.

**Comment by HaiyangDING** [comment]
CC @HaiyangDING

**Comment by dalanlan** [comment]
I could be wrong, yet i was thinking [this line](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/algorithm/predicates/predicates.go#L155) and also [this](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/algorithm/predicates/predicates.go#L161~L164) aim to manage over commit issue. Could you please kindly clarify this for me?

**Comment by dalanlan** [comment]
Plus, what i mainly concern is that scheduler uses resource requested other than usage currently. 
Yet some users may not specify those labels at all , in which case the resource consumed by the those pods are completely ignored currently (correct me if i'm wrong). This makes (2) reasonable.

**Comment by HaiyangDING** [comment]
@dalanlan @davidopp 

Just clarify something that I understand, correct me if wrong.

By default, the resource request defaults to resource limit if resource limit is specified. However, if resource limit is not set, then resource request would not be set either.

This is why we need [limit-range](https://github.com/kubernetes/kubernetes/tree/master/docs/admin/limitrange). Unfortunately, this doc is a bit out-of-date and I have a PR to update this at #13724. Basically, if you set a limit-range, then the resource request/limit would have their default value if not specified by the user.

**Comment by davidopp** [comment]
I'm sorry, you are correct, #1 is already implemented (#5547).

For (2) we do have something a little like this -- LeastRequestedPriority pretends that Pods that request zero CPU actually request 0.1 CPU and Pods that request zero memory actually request 200 MB of memory. In the case where the user requests no resources for any Pods, it should be equivalent to spreading based on number of running Pods. In the case where there is a mixture of Pods that request resources and Pods that don't, it is not exactly the same as spreading based on number of running Pods, but I am not sure it is worth implementing explicit spreading based on number of running pods.

The real solution for the problem you're talking about is to take resource usage on the machine into account when scheduling. We're planning to do that soon. I'll search for the issue number.

**Comment by davidopp** [comment]
It's #12455.

**Comment by dalanlan** [comment]
Yeah i'm aware of that :)
It's blocked by other pr for now, so i was thinking we could have some workaround.

**Comment by davidopp** [comment]
I'm hoping we can unblock it soon. Let me check and get back to you.
